src/testing/alerts.rs

```diff
@@ -0,0 +1,525 @@
+                }
+                debug!("Remaining items in this segment: {:?}.", segment.expected);
+            }
+        }
+        exit_alerter
+            .send(())
+            .expect("exit channel shouldn't be closed");
+    }
+
+    async fn run(self, run_as: NodeIndex) {
+        let keychain = self.keychain(run_as).clone();
+        let mut timeout = Delay::new(Duration::from_millis(500)).fuse();
+        futures::select! {
+            _ = self.test(keychain).fuse() => {},
+            _ = timeout => {
+                panic!("Alerter took too long to emit expected items.");
+            },
+        }
+    }
+}
+
+#[tokio::test]
+async fn distributes_alert_from_units() {
+    let n_members = NodeCount(7);
+    let own_index = NodeIndex(0);
+    let forker = NodeIndex(6);
+    let mut test_case = TestCase::new(n_members);
+    let alert = test_case.alert(own_index, test_case.fork_proof(forker, 0).await);
+    let signed_alert = test_case.unchecked_signed(alert.clone(), own_index).await;
+    test_case
+        .incoming_alert(alert.clone())
+        .outgoing_message(AlertMessage::ForkAlert(signed_alert), Recipient::Everyone);
+    test_case.run(own_index).await;
+}
+
+#[tokio::test]
+async fn reacts_to_correctly_incoming_alert() {
+    let n_members = NodeCount(7);
+    let own_index = NodeIndex(0);
+    let alerter_index = NodeIndex(1);
+    let forker = NodeIndex(6);
+    let mut test_case = TestCase::new(n_members);
+    let fork_proof = test_case.fork_proof(forker, 0).await;
+    let alert = test_case.alert(alerter_index, fork_proof.clone());
+    let signed_alert_hash = test_case
+        .indexed_unchecked_signed(Signable::hash(&alert), own_index)
+        .await;
+    let signed_alert = test_case
+        .unchecked_signed(alert.clone(), alerter_index)
+        .await;
+    test_case
+        .incoming_message(AlertMessage::ForkAlert(signed_alert))
+        .outgoing_notification(ForkingNotification::Forker(fork_proof));
+    for i in 1..n_members.0 {
+        test_case.outgoing_message(
+            AlertMessage::RmcMessage(own_index, RmcMessage::SignedHash(signed_alert_hash.clone())),
+            Recipient::Node(NodeIndex(i)),
+        );
+    }
+    test_case.run(own_index).await;
+}
+
+#[tokio::test]
+async fn notifies_about_finished_alert() {
+    let n_members = NodeCount(7);
+    let own_index = NodeIndex(0);
+    let alerter_index = NodeIndex(1);
+    let forker = NodeIndex(6);
+    let mut test_case = TestCase::new(n_members);
+    let fork_proof = test_case.fork_proof(forker, 0).await;
+    let alert = test_case.alert(alerter_index, fork_proof.clone());
+    let alert_hash = Signable::hash(&alert);
+    let signed_alert = test_case
+        .unchecked_signed(alert.clone(), alerter_index)
+        .await;
+    test_case
+        .incoming_message(AlertMessage::ForkAlert(signed_alert))
+        .outgoing_notification(ForkingNotification::Forker(fork_proof))
+        .wait();
+    for i in 1..n_members.0 - 1 {
```

Comment:
Here we iterate from 1 to 5. Would iterating from 1 to 4 suffice? together with the own index 0 we would have > 2*N/3 honest nodes

Reply:
True, but I'm trying not to make these tests too specific – this is "all honest nodes respond". Should I make another one that checks the minimum? It feels more like a problem for `signed` testing, but I'm not sure.

Reply:
This is fine in the current shape, leave it as is